### ISSUE TITLE
Improve error message for unsupported shells

### ISSUE THREAD

**Issue by avibrazil** [cockpit-project/cockpit#20741]
### Explain what happens

1. Change user shell to /bin/xonsh or /usr/bin/xonsh
2. Try to login with this user; you'll get permission denied

System logs (journalctl) shows successful login but then end of session. See below.

Xonsh is a Python-powered Unix shell and it was installed from Fedora's packages:
```
# rpm -qi xonsh
Name        : xonsh
Version     : 0.14.1
Release     : 2.fc40
Architecture: noarch
Install Date: Wed 26 Jun 2024 04:52:38 PM -03
Group       : Unspecified
Size        : 6356150
License     : BSD-2-Clause and MIT
Signature   : RSA/SHA256, Sat 27 Jan 2024 12:54:42 PM -03, Key ID 0727707ea15b79cc
Source RPM  : xonsh-0.14.1-2.fc40.src.rpm
Build Date  : Sat 27 Jan 2024 07:00:30 AM -03
Build Host  : buildvm-s390x-20.s390.fedoraproject.org
Packager    : Fedora Project
Vendor      : Fedora Project
URL         : https://xon.sh
Bug URL     : https://bugz.fedoraproject.org/xonsh
Summary     : A general purpose, Python-ish shell
Description :
xonsh is a Python-powered, cross-platform, Unix-gazing shell language and
command prompt.

The language is a superset of Python 3.6+ with additional shell primitives.
xonsh (pronounced conch) is meant for the daily use of experts and novices
alike.
```

### Version of Cockpit

320

### Where is the problem in Cockpit?

None

### Server operating system

Fedora

### Server operating system version

40

### What browsers are you using?

Firefox, Safari macOS, Safari on iPhone, Safari on iPad

### System log

```shell
Jul 10 20:21:35 myhost.com systemd[1]: Started session-77.scope - Session 77 of User myuser.
Jul 10 20:21:35 myhost.com cockpit-session[25398]: pam_unix(cockpit:session): session opened for user myuser(uid=1000) by myuser(uid=0)
Jul 10 20:21:35 myhost.com systemd[1]: session-77.scope: Deactivated successfully.
Jul 10 20:21:35 myhost.com systemd-logind[1991]: Session 77 logged out. Waiting for processes to exit.
Jul 10 20:21:35 myhost.com systemd-logind[1991]: Removed session 77.
```

**Comment by martinpitt** [comment]
Cockpit makes *one* assumption about the login shell: `exit 71` exits the shell with code 71. I take it `xonsh` doesn't do that?

Sorry, but Cockpit has to do some very minimal check if the user's login shell is valid (as opposed to e.g. `/bin/false` or `/sbin/nonexisting`). This means that you can't use Cockpit with a shell like that.

**Comment by avibrazil** [comment]
Indeed doesn’t do that, just checked.
Why cockpit has such an exquisite and picturesque requirement about the user shell? I’m ready to request xonsh to do that via a bug report, but what is the reason I should put there?

Thank you in advance

**Comment by avibrazil** [comment]
I’d also recommend and suggest to change the error message in the UI to something like “unsupported shell” or be more verbose and specific in the logs.

It took me ages to discover that the problem is related to the default user shell, since I changed the login shell a long time ago, using it successfully, but use cockpit only occasionally for administrative tasks.

**Comment by martinpitt** [comment]
OK, reopening to look into improving the error message.

The problem is that it shouldn't be possible to log into Cockpit for (hand-wavy) "disabled/invalid" shells, such as system users with /bin/false or /sbin/nologin. We need *some* way to determine whether that's the case. Actual shell logins don't have that problem as `false` or `nologin` will just immediately exit, but cockpit isn't a shell program -- we need to start something else as session leader (cockpit-bridge). We even started that *through* the shell in the past, but that ran into the very same problem -- there isn't enough common ground/syntax between all possible shells to do e.g. input redirection. We also tried to read `/etc/shells`, but it's also too inconsistent and led to problems. So the third iteration was the `exit 71` trick which gives the fewest problems overall.

This is a really tough problem unfortunately.

**Comment by avibrazil** [comment]
Ok, much clearer now.

Just to add that Xonsh is in /etc/shells and I think it will behave as expected by Cockpit. But not the exit 71 hack for now.

**Comment by supakeen** [comment]
@martinpitt Would it make sense to instead of `exit 71` (which is not POSIX) use say `echo 71` or one of the other POSIX required commands for shells? :)

**Comment by martinpitt** [comment]
@supakeen it'd be much harder, as you need to do input redirection, string allocation, etc. -- and all of that from a minimal suid root wrapper; and reading input from a program is by itself hard in C. So that doesn't fill me with joy either..

**Comment by martinpitt** [comment]
Fixed in PR #22244 with screenshots. @avibrazil @supakeen  let me know what you think. Thanks!

**Comment by avibrazil** [comment]
When this is coming to official Fedora 42? This is when I'll have the chance to test it.

**Comment by martinpitt** [comment]
@avibrazil We just had a release two days ago, so in two weeks. But in the meantime you can test the COPR if you want to: https://dashboard.packit.dev/jobs/copr/2641037

**Comment by PaddiM8** [comment]
I am not quite sure I understand what it does. If I do `exit 71` in my shell, it does exit and set the exit code to 71. 

<img width="417" height="380" alt="Image" src="https://github.com/user-attachments/assets/96f244d7-178d-4045-b785-996056107c6e" />

How does it check this?

**Comment by jelly** [comment]
> I am not quite sure I understand what it does. If I do `exit 71` in my shell, it does exit and set the exit code to 71.
> <img alt="Image" width="417" height="380" src="https://private-user-images.githubusercontent.com/19822996/526230188-96f244d7-178d-4045-b785-996056107c6e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjU3ODgzMjcsIm5iZiI6MTc2NTc4ODAyNywicGF0aCI6Ii8xOTgyMjk5Ni81MjYyMzAxODgtOTZmMjQ0ZDctMTc4ZC00MDQ1LWI3ODUtOTk2MDU2MTA3YzZlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjE1VDA4NDAyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI0NjE5ODhjNjE2OWVlZDRhZWU2MzA2OTAwZGZhYmU2NDA3ZDNmZTRhNTExODQ0NzYyMTE2NmY4NGY1YmRkMTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.hzQfR3tNZqXAkEb5EeTGRUbUqmRri3y7xyDBQmQBK2Y">
> 
> How does it check this?

What is your exact question? You have `elk`  as your default shell and Cockpit doesn't work?

**Comment by PaddiM8** [comment]
> What is your exact question? You have elk as your default shell and Cockpit doesn't work?

Yes. I have the same issue as OP but the exit command does make it exit with that exit code
